Lib/warnings.py

```diff
@@ -269,22 +269,33 @@ def _getcategory(category):
     return cat
 
 
+def _is_internal_filename(filename):
+    return 'importlib' in filename and '_bootstrap' in filename
+
+
+def _is_filename_to_skip(filename, skip_file_prefixes):
+    return any(filename.startswith(prefix) for prefix in skip_file_prefixes)
+
+
 def _is_internal_frame(frame):
     """Signal whether the frame is an internal CPython implementation detail."""
-    filename = frame.f_code.co_filename
-    return 'importlib' in filename and '_bootstrap' in filename
+    return _is_internal_filename(frame.f_code.co_filename)
 
 
-def _next_external_frame(frame):
-    """Find the next frame that doesn't involve CPython internals."""
+def _next_external_frame(frame, skip_file_prefixes):
+    """Find the next frame that doesn't involve Python or user internals."""
     frame = frame.f_back
-    while frame is not None and _is_internal_frame(frame):
+    while frame is not None and (
+            _is_internal_filename(filename := frame.f_code.co_filename) or
+            _is_filename_to_skip(filename, skip_file_prefixes)):
         frame = frame.f_back
     return frame
 
 
 # Code typically replaced by _warnings
-def warn(message, category=None, stacklevel=1, source=None):
+def warn(message, category=None, stacklevel=1, source=None,
+         *,
+         skip_file_prefixes: tuple[str, ...] = ()):
```

Comment:
The stlidb used to proscribe annotations, now a few functions have them.
Here adding the annotation may help a little  but not as much as a full docstring, for example to say whether the «prefixes» are true path prefixes or sub-strings (e.g. does `/usr/lib` match `/usr/lib/spam.py` and `/usr/lib64/spam.py`).
Also, maybe it would be more common to use `None` for default value.

(Very minor note: using a whole line just for `*,` marker is not usual)

Reply:
annotation removed. it wasn't going to add value here. this code isn't actually used by CPython, the C extension module is.